File: src/task/mutex.cpp
/**
 * @copyright Copyright The SimpleKernel Contributors
 */

#include "mutex.hpp"

#include "kernel.h"
#include "kernel_log.hpp"
#include "task_manager.hpp"

auto Mutex::Lock() -> bool {
  auto current_task = TaskManagerSingleton::instance().GetCurrentTask();
  if (current_task == nullptr) {
    klog::Err("Mutex::Lock: Cannot lock mutex '%s' outside task context",
              name_);
    return false;
  }

  Pid current_pid = current_task->pid;

  // 检查是否递归获取锁
  if (IsLockedByCurrentTask()) {
    klog::Warn("Mutex::Lock: Task %d tried to recursively lock mutex '%s'",
               current_pid, name_);
    return false;
  }

  // 尝试获取锁
  bool expected = false;
  while (!locked_.compare_exchange_weak(
      expected, true, std::memory_order_acquire, std::memory_order_relaxed)) {
    // 锁被占用，阻塞当前任务
    klog::Debug("Mutex::Lock: Task %d blocking on mutex '%s'", current_pid,
                name_);
    TaskManagerSingleton::instance().Block(resource_id_);

    // 被唤醒后重新尝试
    expected = false;
  }

  // 成功获取锁，记录所有者
  owner_.store(current_pid, std::memory_order_release);
  klog::Debug("Mutex::Lock: Task %d acquired mutex '%s'", current_pid, name_);
  return true;
}

auto Mutex::UnLock() -> bool {
  auto current_task = TaskManagerSingleton::instance().GetCurrentTask();
  if (current_task == nullptr) {
    klog::Err("Mutex::UnLock: Cannot unlock mutex '%s' outside task context",
              name_);
    return false;
  }

  Pid current_pid = current_task->pid;

  // 检查是否由持有锁的任务释放
  if (!IsLockedByCurrentTask()) {
    klog::Warn(
        "Mutex::UnLock: Task %d tried to unlock mutex '%s' it doesn't own",
        current_pid, name_);
    return false;
  }

  // 释放锁
  owner_.store(std::numeric_limits<Pid>::max(), std::memory_order_release);
  locked_.store(false, std::memory_order_release);

  klog::Debug("Mutex::UnLock: Task %d released mutex '%s'", current_pid, name_);

  // 唤醒等待此锁的任务
  TaskManagerSingleton::instance().Wakeup(resource_id_);

  return true;
}

auto Mutex::TryLock() -> bool {
  auto current_task = TaskManagerSingleton::instance().GetCurrentTask();
  if (current_task == nullptr) {
    klog::Err("Mutex::TryLock: Cannot trylock mutex '%s' outside task context",
              name_);
    return false;
  }

  Pid current_pid = current_task->pid;

  // 检查是否递归获取锁
  if (IsLockedByCurrentTask()) {
    klog::Debug(
        "Mutex::TryLock: Task %d tried to recursively trylock mutex '%s'",
        current_pid, name_);
    return false;
  }

  // 尝试获取锁（非阻塞）
  bool expected = false;
  if (locked_.compare_exchange_strong(expected, true, std::memory_order_acquire,
                                      std::memory_order_relaxed)) {
    // 成功获取锁，记录所有者
    owner_.store(current_pid, std::memory_order_release);
    klog::Debug("Mutex::TryLock: Task %d acquired mutex '%s'", current_pid,
                name_);
    return true;
  }

  // 锁被占用
  klog::Debug("Mutex::TryLock: Task %d failed to acquire mutex '%s'",
              current_pid, name_);
  return false;
}

auto Mutex::IsLockedByCurrentTask() const -> bool {
  auto current_task = TaskManagerSingleton::instance().GetCurrentTask();
  if (current_task == nullptr) {
    return false;
  }

  return locked_.load(std::memory_order_acquire) &&
         owner_.load(std::memory_order_acquire) == current_task->pid;
}

File: src/task/mutex.rs
//! 内核阻塞互斥锁——竞争时任务进入 Blocked 状态，而非自旋等待。

#[cfg(target_os = "none")]
mod inner {
    use core::sync::atomic::{AtomicU64, AtomicUsize, Ordering};

    use crate::task::resource_id::ResourceId;

    const NO_OWNER: usize = usize::MAX;

    /// 内核阻塞互斥锁
    ///
    /// 竞争时任务进入 Blocked 状态（让出 CPU），而非自旋等待。
    /// 适用于可能长时间持有的锁。
    pub struct KMutex {
        /// 互斥锁 ID（用于 ResourceId）
        id: u64,
        /// 当前持有者 PID（NO_OWNER 表示未持有）
        owner: AtomicUsize,
    }

    /// KMutex ID 计数器
    static NEXT_MUTEX_ID: AtomicU64 = AtomicU64::new(1);

    impl KMutex {
        /// 创建新的阻塞互斥锁。
        pub fn new() -> Self {
            Self {
                id: NEXT_MUTEX_ID.fetch_add(1, Ordering::Relaxed),
                owner: AtomicUsize::new(NO_OWNER),
            }
        }

        /// 获取锁——如果锁已被持有，当前任务进入 Blocked 状态。
        pub fn lock(&self) {
            loop {
                let current_pid = crate::task::current_task().pid();
                match self.owner.compare_exchange(
                    NO_OWNER,
                    current_pid,
                    Ordering::Acquire,
                    Ordering::Relaxed,
                ) {
                    Ok(_) => return,
                    Err(owner) => {
                        if owner == current_pid {
                            panic!("KMutex: recursive lock by pid={}", current_pid);
                        }
                        crate::task::block_on(ResourceId::Mutex(self.id));
                    }
                }
            }
        }

        /// 释放锁——唤醒一个等待任务。
        pub fn unlock(&self) {
            self.owner.store(NO_OWNER, Ordering::Release);
            crate::task::wakeup_one(ResourceId::Mutex(self.id));
        }
    }
}

#[cfg(target_os = "none")]
pub use inner::KMutex;
